81-question-mark-chrysalis/technical_specification.md

```diff
@@ -0,0 +1,263 @@
+
+data structure:
+
+```python
+class Route(BaseModel):
+   name: str | None
+   input_schema_name: str
+   output_schema_name: str
+   workflow_name: str
+```
+
+`Config`
+
+- Purpose: model the current active transformation configuration version.
+
+data structure:
+
+```python
+
+class BaseConfig(BaseModel):
+   schemas: list[Model]
+   workflows: list[Workflow]
+   routes: list[Route]
+
+
+class Config(BaseModel):
+   version: int = 0  # always 0
+   created: datetime # is the date when the config is validated and activated.
+   config: BaseConfig
+```
+
+
+`AnnotatedEMPack`
+
+Purpose: model incoming and derived annotated EM datapacks that are to be processed or published.
+
+data structure:
+
+```python
+class AnnotatedEMPack(BaseModel):
+   id: uuid
+   schemapack_name: str
+   original_name: str | None
+   datapack: DataPack
+   annotation: dict
+```
+
+#### Implementation Details of Collections
+
+The implementation of the EM transformation service will involve the following key components.
+
+1. `models`
+   
+   Consists of Model objects
+   * `name` is a unique human readable name of the schema also used as an identifier
+   * `description` is a human readable description of the schema
+   * `original` is a boolean flag indicating whether the schema is an original or derived schema
+   * `version` is the version of the schema, initially None if it is not an original schema
+   * `schemapack` is the schema in schemapack format, initially None if it is not an original schema
+   * `publish` is a boolean flag indicating whether the AnnotatedEMPacks conforming to this schema should be published
+   * `order` is the order in a topological ordering of the schemas in the transformation graph, initially None if not yet computed
+
+   This collection is populated via a transformation configuration YAML file.
+
+2. `workflows`
+
+   Consists of Workflow objects
+   * `name` is a unique human readable name for the workflow also used as an identifier. It should be a name indicating the purpose of the workflow for easier debugging and understanding of the operations.
+   * `description` is a more verbose description of the workflow
+   * `workflow` is the definition of the workflow in metldata format
+
+   This collection is populated via a configuration YAML file that contains the defined ghga transformation workflows. 
+
+
+3. `routes`
+
+   Consists of Route object
+   * `name` is a unique human readable name for the workflow route also used as an identifier. It must be calculated from the input and output schema names and the workflow name, e.g. `{input_schema_name}:{workflow_name}:{output_schema_name}`. 
+   * `input_schema_name` is the name/id of the input schema that the workflow route accepts
+   * `output_schema_name` is the name/id of the output schema that the workflow route produces 
+   * `workflow_name` is a workflow name/id that is to be applied on a datapack / schemapack 
+   
```

Comment:
see above: since the fields are redundant, this needs a validator that checks for consistency and derives missing fields if only specified in one way